✨ Add kubectl workspace plugin and adjust usage output

hardys opened the pull request:
## Summary

Elsewhere we describe aliases of ws/workspace/workspaces but the
plugin only works with ws/workspaces atm - this makes it consistent
and allows us to share the same usage with the underlying kcp
plugin subcommand.

## Related issue(s)

Discussed in #687 and may enable some clearer and/or more consistent docs, but doesn't actually fix that issue

ncdc commented on cmd/kubectl-workspace/main.go:
```diff
@@ -0,0 +1,50 @@
+limitations under the License.
+*/
+
+package main
```
symlinks or hardlinks, yes

hardys replied:
Now updated to create the additional plugins via symlink in the Makefile

hardys commented:
> Just FYI, I have a ton of plugin structural changes in #1701, in case you're interested.

Ah thanks, will review! :+1:

hardys commented:
I reviewed #1701 and tried it locally - lgtm but doesn't address the issues covered in the PR so I rebased this and adjusted the `workspace` plugin part to use symlinks instead of further duplicated code (this also removes the current duplication between `kubectl-ws` and `kubectl-workspaces`)

ncdc commented on Makefile:
```diff
@@ -107,6 +107,9 @@ build-kind-images: build-kind-images-ko
 install: WHAT ?= ./cmd/...
 install:
 	go install -ldflags="$(LDFLAGS)" $(WHAT)
+	$(eval KUBECTL_WORKSPACES=$(shell which kubectl-workspaces))
```
Don't have to do it in this PR, but since the `Use` field is `workspace` and the package is `pkg/cliplugins/workspace` (both singular), should we consider renaming `cmd/kubectl-workspaces` to `cmd/kubectl-workspace` and then having `kubectl-workspaces` be a symlink to `kubectl-workspaces`?

stevekuznetsov replied:
This assumes that `${GOPATH}/bin` is in the user's `${PATH}` which may not be true (is not true for me, for instance). We could look at `${GOBIN}` and the defaults explicitly, perhaps:

> Executables are installed in the directory named by the GOBIN environment
variable, which defaults to $GOPATH/bin or $HOME/go/bin if the GOPATH
environment variable is not set. Executables in $GOROOT
are installed in $GOROOT/bin or $GOTOOLDIR instead of $GOBIN.

hardys replied:
> This assumes that ${GOPATH}/bin is in the user's ${PATH} which may not be true 

Ack, I made this assumption because the [kubectl plugin docs](https://kubernetes.io/docs/tasks/extend-kubectl/kubectl-plugins/) say plugins are discovered via `$PATH`

I did also consider parsing the `go env` output, but as you say that is complicated by the fact we have to detect whether to use `$GOBIN` or `$GOPATH/bin` (GOBIN isn't actually defaulted in the `go env` output even if it is internally for `go install` ...), so after looking at the plugin docs I decided to use `which` instead.

I'll see if there's a clean way to handle this case, thanks for the feedback!

sttts replied:
+1 for GOBIN. Build path does not have to equal the users' kubectl plugin in use and hence in the PATH.

hardys replied:
Sorry for the slow update, this now handles GOBIN and I moved the cmd directory to workspace (singular) as requested by @ncdc - thanks for the review!

hardys commented:
Ready for another review, previous comments addressed - thanks for the feedback!

One thing to mention - this only symlinks in the `install` Makefile target - this means we won't have the CLI aliases in the container image, because the [Dockerfile](https://github.com/kcp-dev/kcp/blob/main/Dockerfile#L52) uses the `build` target then copies `kubectl-*`

Do we want/need the ws/workspaces aliases in the Dockerfile (or build target) to resolve that?

hardys commented:
/retest

ncdc commented:
goreleaser config needs updating

ncdc commented:
Re the Dockerfile, we should probably ensure the symlinks get into the image. Would it be possible to have them created in the `build` target?

hardys commented:
> Re the Dockerfile, we should probably ensure the symlinks get into the image. Would it be possible to have them created in the `build` target?

Yes this seems like the best option given that COPY resolves any symlinks created by the build target in the makefile - now updated

hardys commented:
> goreleaser config needs updating

Ack now updated to reflect the new binary naming, although looking at the docs it seems like we can [only create symlinks when generating OS packages](https://goreleaser.com/customization/nfpm/) - I'll double check this though as I'm not very familiar with goreleaser.

hardys commented:
@ncdc I updated the goreleaser config - AFAICS to generate symlinks in the archive I need to create them in a post build hook, then copy the resulting files, this seems to work running locally but happy to hear if there is a cleaner way :) 

```
$ tar -tvf dist/kubectl-kcp-plugin_0.1.0-SNAPSHOT-c6d4c591_linux_amd64.tar.gz
lrwxrwxrwx shardy/shardy     0 2022-09-26 15:50 bin/kubectl-workspaces -> kubectl-workspace
lrwxrwxrwx shardy/shardy     0 2022-09-26 15:50 bin/kubectl-ws -> kubectl-workspace
-rwxrwxr-x shardy/shardy 94323912 2022-09-26 15:50 bin/kubectl-workspace
-rwxrwxr-x shardy/shardy 101777416 2022-09-26 15:50 bin/kubectl-kcp
```

hardys commented:
@ncdc this is ready for another review pass when you get a moment, thanks!
